File: C++/Queue/QueueUsingArray.cpp
/*	
	@author
	Aakash.Verma

	Queue is a FIFO/LILO data structures:
	FIFO stands for First In First Out 
	
	We can't access any element except front. Front is that element which has been inserted at first
	and will be retrieved first. Rear is that element which is inserted at last and will be deleted 
	at last.

	I am assuming that you are here after going through the theory of Queue and looking for implementation.

	Output: 	

	Enqueue Operation Done, rear is at  11
	Enqueue Operation Done, rear is at  12
	Enqueue Operation Done, rear is at  13
	Deleted Element is:  11
	Front element is:  12

	Compile using:  g++ filename.cpp
	Run using: ./a.out filename.cpp 
*/


#include <iostream>
using namespace std;

  
const int SIZE = 10; 
int front = -1;
int rear = -1;
int queue[SIZE];

/* Checking if the Queue is full */
bool isFull() {
	if (front == 0 && rear == SIZE - 1)
	  	return true;
	if (front == rear + 1)
	  	return true;
	return false;
}

/* Check if the Queue is empty */
bool isEmpty() {
	if (front == -1)
	  	return true;
	else
	  	return false;
}

// Adding an element
void enQueue(int element) {
	if (isFull()) {
	  	cout << "Queue is full" << endl;
	} 
	else {
	  	if (front == -1)
	    	front = 0;
	  	rear = (rear + 1) % SIZE;
	  	queue[rear] = element;
	  	cout << "Enqueue Operation Done, rear is at " << element << endl;
	}
}

/* Deleting an element */
int deQueue() {
	int element;
	if (isEmpty()) {
	  	cout << "Queue is empty" << endl;
	  	return -1;
	} 
	else {
	 	element = queue[front];
	  	if (front == rear) {
	    	front = -1;
	    	rear = -1;
	  	} /* Q has only one element, so we reset the queue after deleting it. */
	  	else {
	    	front = (front + 1) % SIZE;
	  	}
	  	return element;
	}
}

int peek() {
	if (isEmpty()) {
	  	cout << "Queue is empty." << endl;
	  	return -1;
	} 
	else {
	 	return queue[front];
	}
}

int main(int argc, char const *argv[]) {
    enQueue(11);
    enQueue(12);
    enQueue(13);
    int deleted = deQueue();
    cout << "Deleted Element is: " << deleted << endl;
    cout << "Front Element is: " << peek() << endl;
    
	return 0;
}


	    

File: Python/Queue/QueueUsingArray.py
# @author
# Aakash.Verma

# Queue is a FIFO/LILO data structures:
# FIFO stands for First In First Out 

# We can't access any element except front. Front is that element which has been inserted at first
# and will be retrieved first. Rear is that element which is inserted at last and will be deleted 
# at last.

# I am assuming that you are here after going through the theory of Queue and looking for implementation.

# Output: 

# Enqueue Operation Done, rear is at  11
# Enqueue Operation Done, rear is at  12
# Enqueue Operation Done, rear is at  13
# Deleted Element is:  11
# Front element is:  12

# Run using: python3 QueueUsingArray.py

class QueueUsingArray:
    # We are assuming that the size of our queue is 10, you may take any size
    SIZE = 10

    queue = []  # queue

    # As soon as the object of this class is created, front & rear is assigned as -1.
    def __init__(self):
        self.front = -1
        self.rear = -1

    # It checks whether the queue is empty or not, obviously if front is -1, the queue will be empty.
    def isEmpty(self):
        if (self.front == -1):
            return True
        return False

    # It checks whether the queue is full or not, for better explanation watch the video on YouTube (Innoskrit)
    def isFull(self):
        if (self.front == 0 and self.rear == self.SIZE - 1):
            return True
        if (self.front == self.rear + 1):
            return True
        return False

    def enQueue(self, element):
        if (self.isFull()):
            print("Queue is full")
        else:
            if self.front == -1:
                self.front = 0
            self.rear = (self.rear + 1) % self.SIZE
            self.queue.append(element)
            print("Enqueue Operation Done, rear is at ", element)

    def deQueue(self):
        if (self.isEmpty()):
            print("Queue is empty")
            return -1
        else:
            element = self.queue[self.front]

            if (self.front == self.rear):
                self.front = -1
                self.rear = -1
            # Q has only one element, so we reset the queue after deleting it.

            else:
                self.front = (self.front + 1) % self.SIZE
            return element

    def peek(self):
    	if(self.isEmpty()):
    		print("Queue is empty")
    		return -1
    	else :
    		return self.queue[self.front]


if __name__ == '__main__':

    q = QueueUsingArray()
    q.enQueue(11)
    q.enQueue(12)
    q.enQueue(13)
	deleted = q.dequeue()
	print("Deleted Element is: ", deleted)
	print("Front element is: ", q.peek())



